aptos-move/block-executor/src/code_cache_global.rs

```diff
@@ -160,131 +155,167 @@ where
                     ));
                 } else {
                     // Otherwise, remove the invalid entry.
+                    let size = entry.get().module_code().extension().size_in_bytes();
+                    self.size.fetch_sub(size, Ordering::Relaxed);
                     entry.remove();
                 }
             }
 
             if module.code().is_verified() {
-                let mut module = module.as_ref().clone();
-                module.set_version(None);
-                let prev =
-                    module_cache.insert(key.clone(), ImmutableModuleCode::new(Arc::new(module))?);
+                self.size
+                    .fetch_add(module.extension().size_in_bytes(), Ordering::Relaxed);
+                let entry =
+                    Entry::new(module).expect("Module has been checked and must be verified");
+                let prev = module_cache.insert(key.clone(), entry);
 
                 // At this point, we must have removed the entry, or returned a panic error.
                 assert!(prev.is_none())
             }
         }
-
-        if module_cache.len() > self.capacity {
-            module_cache.clear();
-        }
-
         Ok(())
     }
 
     /// Insert the module to cache. Used for tests only.
     #[cfg(any(test, feature = "testing"))]
-    pub fn insert(&self, key: K, module: Arc<ModuleCode<DC, VC, E, Option<TxnIndex>>>) {
-        self.module_cache
-            .acquire()
-            .insert(key, ImmutableModuleCode::new(module).unwrap());
+    pub fn insert(&self, key: K, module: Arc<ModuleCode<DC, VC, E>>) {
+        self.size
+            .fetch_add(module.extension().size_in_bytes(), Ordering::Relaxed);
+        self.module_cache.acquire().insert(
+            key,
+            Entry::new(module).expect("Module code should be verified"),
+        );
     }
 
     /// Removes the module from cache. Used for tests only.
     #[cfg(any(test, feature = "testing"))]
     pub fn remove(&self, key: &K) {
```

Comment:
hm wouldn't it be useful for tests to return true/false? (or Option/Some(prev) if we wanted to go that far)